data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -340,7 +355,11 @@ public void nonNullWrite(int rowId, BigDecimal data, ColumnVector output) {
 
     @Override
     public void nonNullWrite(int rowId, BigDecimal data, ColumnVector output) {
-      // TODO: validate precision and scale from schema
+      Preconditions.checkArgument(data.scale() == scale,
+          "Cannot write value as decimal(%s,%s), wrong scale: %s", precision, scale, data);
+      Preconditions.checkArgument(data.precision() <= precision && precision <= 38,
```

Comment:
Nit: `precision <= 38` check can be moved into the constructor

Reply:
ditto.